### PR TITLE
[processor/k8sattributes] Add `container.image.repo_digests`

### PR DESCRIPTION
**Pull request by evantorrie**
**Description:**

Adds support for extracting canonical repo digest image references for a container from k8s attributes

**Link to tracking Issue:** #34029 

**Testing:**

1. Adds unit tests for extracting the k8s `.imageID` field from container status API if it is of the `reference.Canonical` form, and storing it in a `ImageRepoDigest` string field. (Note: k8s will only ever return one value here, although the semantic conventions for `container.image.repo_digests` define it to be a `Slice`.  We convert into a Slice lazily when we populate the pcommon data structure.
2. Adds unit tests for including `container.image.repo_digests` in rules for extraction.
3. Includes `container.image.repo_digests` in the e2e tests for k8sattributes processor.

**Documentation:**

Added `container.image.repo_digests` as an example of attributes that can be extracted from a container

### CODE REVIEW

**Comment by evantorrie** [comment]
Implementation notes:
1. This is the first k8sattributes e2e test which returns a `Slice` attribute value. I have not changed the "expectedValue" matcher to explicitly handle slices.  Since there is only ever going to be *one* value in the `container.image.repo_digests` slice attribute, the existing test code which checks for the "expected value" appearing anywhere in "value.ToString()", this will work without having to radically rework the matchers.
2. Although the k8s API structure returns this value in `ImageID`, when we process this into our local data structure for later usage, we use a `ImageRepoDigest` string field - and populate it only if the `ImageID` field returned by the k8s API is of the correct `distribution/reference.Canonical` type.

**Comment by ChrsMark** [comment]
So eventually we are not going to follow the

> I think one way to proceed is to also use the github.com/distribution/reference package to attempt to parse the output of .imageID from the k8s API response. If it is a canonical reference, then populate the [semantic attribute](https://opentelemetry.io/docs/specs/semconv/attributes-registry/container/#container-attributes) container.image.repo_digests with those values (of which it seems there will only be one from the existing CRI implementations (?)). If it is not, then populate container.image.id.

from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32314#issuecomment-2207580404?

Doesn't this fallback to `container.image.id` makes sense in case we cannot extract the repo digests?

**Comment by evantorrie** [comment]
> Doesn't this fallback to `container.image.id` makes sense in case we cannot extract the repo digests?

I was trying to simplify this PR by limiting the scope to just populating `container.image.repo_digests` if what the CRI returns is indeed a valid repo digest.

Including `container.image.id` could be done in a subsequent PR (I do have a version of this code which does the fallback), although I still am unclear on the use case for someone wanting to know `container.image.id` in cases where it is not a valid repo reference, and instead is just a sha256 ref that is valid only for the local CRI used by the k8s apiserver.

Yes, for theoretical completeness sake, it seems to make sense, but in practice, I don't know whether someone really would use this. Maybe my imagination (or understanding of what value a sha256 ref provides) is too limited.

**Comment by ChrsMark** [comment]
@evantorrie sounds good!

>  Yes, for theoretical completeness sake, it seems to make sense, but in practice, I don't know whether someone really would use this. Maybe my imagination (or understanding of what value a sha256 ref provides) is too limited.

The only use-case I would see here is if somebody wants to reference an image even if that's not coming from a repo. For a security analysis, generic issue investigation or sth. But yes it can be done later.

**Comment by evantorrie** [comment]
Unrelated unit test failing due to this:  https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33998

File: processor/k8sattributesprocessor/options.go
Comment (ChrsMark) [inline processor/k8sattributesprocessor/options.go:27]: 👍🏽

File: processor/k8sattributesprocessor/README.md
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:78]: I wonder if this could be a bit more explanatory 🤔

**Review by ChrsMark — COMMENTED** [review]
The change looks good to me based on the prior discussions within the various related issues. Codeowners of the processor should also review/confirm.

File: processor/k8sattributesprocessor/README.md
Comment (evantorrie) [reply processor/k8sattributesprocessor/README.md:78]: Maybe "if k8s CRI populates repository + digest".
But then we need to explain what CRI means to people who don't really know/care about that.
Maybe just a link to the explanation in the semantic conventions? https://github.com/open-telemetry/semantic-conventions/blob/f79fe6b61438bfa0804df9c254c402fd0d71f42a/model/registry/container.yaml#L60-L71

Comment (ChrsMark) [reply processor/k8sattributesprocessor/README.md:78]: Would adding both work here? I think it's better to explicitly surface this technical detail, even if that might be unknown to some users (they can always learn about it) rather than hiding it.

Comment (evantorrie) [reply processor/k8sattributesprocessor/README.md]: Added link to semantic conventions registry and enhanced comment about how repo_digests field will get populated.

**Comment by evantorrie** [comment]
@open-telemetry/collector-contrib-approvers Could we get another pair of eyes on this pull request? Thanks!

File: processor/k8sattributesprocessor/README.md
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:78]:
```suggestion
   - container.image.repo_digests (if k8s CRI populates [repository digest field](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/container.yaml#L60-L71))
```

File: processor/k8sattributesprocessor/README.md
Comment (ChrsMark) [inline processor/k8sattributesprocessor/README.md:82]:
```suggestion
   - container.image.repo_digests (if k8s CRI populates [repository digest field](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/container.yaml#L60-L71))
```

**Review by ChrsMark — APPROVED** [review]
LGTM

**Comment by ChrsMark** [comment]
@evantorrie the CI failures look unrelated. Consider rebasing as well.

**Comment by evantorrie** [comment]
@TylerHelmuth @fatsheep9146 This has been pretty much complete for a week or so now.  Any chance of getting another review on it from the code-owners?

File: processor/k8sattributesprocessor/processor.go
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:253]: We always set it to a one-item slice. Is it possible to get more of them in the future? Why does the attribute have to be a slice not string?

**Review by dmitryax — APPROVED** [review]
LGTM. Just one question

File: processor/k8sattributesprocessor/processor.go
Comment (evantorrie) [reply processor/k8sattributesprocessor/processor.go:253]: The [semantic conventions define](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/container.yaml#L60-L71) `container.image.repo_digests` to be an array of strings, so even if there's only one of them now, I believe it has to be put into a slice to conform with semconv.

**Pull request merged**
